builtin/common/datastructures.lua

```diff
@@ -0,0 +1,317 @@
+local stack_mt
+stack_mt = {
+	__index = {
+		push = function(self, v)
+			self.n = self.n+1
+			self[self.n] = v
+		end,
+		pop = function(self)
+			local v = self[self.n]
+			self[self.n] = nil
+			self.n = self.n-1
+			return v
+		end,
+		is_empty = function(self)
+			return self.n == 0
+		end,
+		size = function(self)
+			return self.n
+		end,
+		clone = function(self, copy_element)
+			local stack
+			if copy_element then
+				stack = {n = self.n, true}
+				for i = 1, self.n do
+					stack[i] = copy_element(self[i])
+				end
+			else
+				stack, n = self:to_table()
+				stack.n = n
+			end
+			setmetatable(stack, stack_mt)
+			return stack
+		end,
+		to_table = function(self)
+			local t = {}
+			for i = 1, self.n do
+				t[i] = self[i]
+			end
+			return t, self.n
+		end,
+		to_string = function(self, value_tostring)
+			if self.n == 0 then
+				return "empty stack"
+			end
+			value_tostring = value_tostring or tostring
+			local t = {}
+			for i = 1, self.n do
+				t[i] = value_tostring(self[i])
+			end
+			return self.n .. " elements; bottom to top: " ..
+				table.concat(t, ", ")
+		end,
+	}
+}
+
+function core.create_stack(data)
+	local stack
+	if type(data) == "table"
+	and data.input then
+		stack = data.input
+		stack.n = data.n or #data.input
+	else
+		-- setting the first element to true makes it ~10 times faster with
+		-- luajit when the stack always contains less or equal to one element
+		stack = {n = 0, true}
+	end
+	setmetatable(stack, stack_mt)
+	return stack
+end
+
+
+local fifo_mt
+fifo_mt = {
+	__index = {
+		add = function(self, v)
+			local n = self.n_in+1
+			self.n_in = n
+			self.sink[n] = v
+		end,
+		take = function(self)
+			local p = self.p_out
+			if p <= self.n_out then
+				local v = self.source[p]
+				self.source[p] = nil
+				self.p_out = p+1
+				return v
+			end
+			-- source is empty, swap it with sink
+			self.source, self.sink = self.sink, self.source
+			self.n_out = self.n_in
+			self.n_in = 0
+			local v = self.source[1]
+			self.source[1] = nil
+			self.p_out = 2
+			return v

```

Comment:
You could swap the uppermost element with the current one and set the uppermost one to `nil` and when you reach the top you can make the source empty from above.

Reply:
This doesn't improve performance, does it?

Reply:
It could maybe save memory.
It was probably a bad suggestion. >_<